libbeat/autodiscover/provider.go

```diff
@@ -0,0 +1,88 @@
+package autodiscover
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/cfgfile"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Provider for autodiscovery
+type Provider interface {
+	cfgfile.Runner
+}
+
+// Registry holds all known autodiscover providers, they must be added to it to enable them for use
+var Registry = NewRegistry()
+
+// ProviderBuilder creates a new provider based on the given config and returns it
+type ProviderBuilder func(bus.Bus, *common.Config) (Provider, error)
+
+// Register of autodiscovery providers
+type registry struct {
+	// Lock to control concurrent read/writes
+	lock sync.RWMutex
+	// A map of provider name to ProviderBuilder.
+	providers map[string]ProviderBuilder
+}
+
+// NewRegistry creates and returns a new Registry
+func NewRegistry() *registry {
+	return &registry{
+		providers: make(map[string]ProviderBuilder, 0),
+	}
+}
+
+// AddProvider registers a new ProviderBuilder
+func (r *registry) AddProvider(name string, provider ProviderBuilder) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if name == "" {
+		return fmt.Errorf("provider name is required")
+	}
+
+	name = strings.ToLower(name)
+
+	_, exists := r.providers[name]
+	if exists {
+		return fmt.Errorf("provider '%s' is already registered", name)
+	}
+
+	if provider == nil {
+		return fmt.Errorf("provider '%s' cannot be registered with a nil factory", name)
+	}
+
+	r.providers[name] = provider
+	logp.Debug(debugK, "Provider registered: %s", name)
+	return nil
+}
+
+// GetProvider returns the provider with the giving name, nil if it doesn't exist
+func (r *registry) GetProvider(name string) ProviderBuilder {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	name = strings.ToLower(name)
```

Comment:
What is the reason you do the `ToLower`?

Reply:
User input sanitizing, `type: Docker` would still work, I think I copied this from somewhere else but I'm not sure :P

Reply:
I would suggest not to sanatise it and force users to do "correct" config files instead. Reducing magic :-)

Reply:
Just found where I copied this from: https://github.com/elastic/beats/blob/354fdd6792b77ed5f099e017caab7c8ec4fa2ead/metricbeat/mb/registry.go#L77, I'm ok with removing it anyway

Reply:
Undecided on this one. Probably easier on our users. @andrewkroh WDYT?

Reply:
As the person who made the registry case-insensitive I like that it accepts any case as input.